src/data_cell/graph_datacell.h

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <unordered_map>
+#include <vector>
+
+#include "algorithm/hnswlib/hnswalg.h"
+#include "common.h"
+#include "graph_interface.h"
+#include "index/index_common_param.h"
+#include "io/basic_io.h"
+#include "vsag/constants.h"
+
+namespace vsag {
+
+/**
+ * built by nn-descent or incremental insertion
+ * add neighbors and pruning
+ * retrieve neighbors
+ */
+template <typename IOTmpl, bool is_adapter>
+class GraphDataCell;
+
+template <typename IOTmpl>
+class GraphDataCell<IOTmpl, false> : public GraphInterface {
+public:
+    GraphDataCell(const nlohmann::json& graph_json_params,
+                  const nlohmann::json& io_json_params,
+                  const IndexCommonParam& common_param);
+
+    void
+    InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) override;
+
+    uint32_t
+    GetNeighborSize(InnerIdType id) override;
+
+    void
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) override;
+
+    inline void
+    SetIO(std::shared_ptr<BasicIO<IOTmpl>> io) {
+        this->io_ = io;
+    }
+
+    /****
+     * prefetch neighbors of a base point with id
+     * @param id of base point
+     * @param neighbor_i index of neighbor, 0 for neighbor size, 1 for first neighbor
+     */
+    void
+    Prefetch(InnerIdType id, uint32_t neighbor_i) override {
+        io_->Prefetch(id * this->code_line_size_ + sizeof(uint32_t) +
+                      neighbor_i * sizeof(InnerIdType));
+    }
+
+    void
+    Serialize(StreamWriter& writer) override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+private:
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
+
+    uint32_t code_line_size_{0};
+};
+
+template <typename IOTmpl>
+GraphDataCell<IOTmpl, false>::GraphDataCell(const nlohmann::json& graph_json_params,
+                                            const nlohmann::json& io_json_params,
+                                            const IndexCommonParam& common_param) {
+    this->io_ = std::make_shared<IOTmpl>(io_json_params, common_param);
+    if (graph_json_params.contains(GRAPH_PARAM_MAX_DEGREE)) {
+        this->maximum_degree_ = graph_json_params[GRAPH_PARAM_MAX_DEGREE];
+    }
+
+    if (graph_json_params.contains(GRAPH_PARAM_INIT_MAX_CAPACITY)) {
+        this->max_capacity_ = graph_json_params[GRAPH_PARAM_INIT_MAX_CAPACITY];
+    }
+
+    this->code_line_size_ = this->maximum_degree_ * sizeof(InnerIdType) + sizeof(uint32_t);
+}
+
+template <typename IOTmpl>
+void
+GraphDataCell<IOTmpl, false>::InsertNeighborsById(InnerIdType id,
+                                                  const Vector<InnerIdType>& neighbor_ids) {
+    if (neighbor_ids.size() > this->maximum_degree_) {
+        logger::error(fmt::format(
+            "insert neighbors count {} more than {}", neighbor_ids.size(), this->maximum_degree_));
```

Comment:
what is the next step after logging errors, return or continue?

Reply:
add comments here about why logging error but not return early

Reply:
I think warning is better